airflow/models.py

```diff
@@ -1957,6 +1832,19 @@ def dag_id(self):
             return 'adhoc_' + self.owner
 
     @property
+    def deps(self):
+        """
+        Returns the list of dependencies for the operator. These differ from execution
+        context dependencies in that they are specific to tasks and can be
+        extended/overriden by subclasses.
```

Comment:
One example might be exponential backoff for the retry period instead of the linear retry, but I think in general these should not be derived to limit complexity. The reason these dependencies are here instead of in execution contexts is because I think they are fundamentally different. E.g. from the infrastructure perspective we don't care if a task wants to remove NotInRetryPeriodDep and retry itself 3 times in quick succession, but if the task uses too many slots in a pool that violates airflow's guarantees and for other tasks that use that pool. There is a flag for contexts that overrides only these dependencies too. I don't feel super strongly about keeping these as separate dependencies since the line is a bit blurry so feel free to push back.